File: Week 2 - Arrays/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Prototypes
int calculate_coleman_liau_index(string input);
void get_text_input_to_calculate(void);

int main(void)
{
    get_text_input_to_calculate();
}

// function to get a user input, call the calculate function
// and return the grade to the user
void get_text_input_to_calculate(void)
{
    string input = get_string("Text: ");
    int index = calculate_coleman_liau_index(input);

    if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

// function that make the Coleman-Liau index calculus
int calculate_coleman_liau_index(string input)
{
    int input_size = strlen(input);
    int sentences = 0;
    int letters = 0;
    int words = 0;
    for (int i = 0; i < input_size; i++)
    {
        if (input[i] == '!' || input[i] == '.' || input[i] == '?')
        {
            sentences++;
        }

        if (!isblank(input[i]) && isalpha(input[i]))
        {
            letters++;
        }

        if (isblank(input[i]))
        {
            words++;
        }

        if (i == input_size - 1)
        {
            words++;
        }
    }

    float l = (letters / (float) words) * 100.0;
    float s = (sentences / (float) words) * 100.0;
    float index = 0.0588 * l - 0.296 * s - 15.8;

    return round(index);
}

File: Week 6 - Python/sentimental-readability/readability.py
def main():
    text = input("Text: ").lower().strip()
    index = calculate_liau_index(text)

    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


def calculate_liau_index(text):
    words = get_words_quantity(text)
    letters = get_letters_quantity(text)
    sentences = get_sentences_quantity(text)

    l = (letters / words) * 100
    s = (sentences / words) * 100
    index = round((0.0588 * l) - (0.296 * s) - 15.8)
    return index


def get_words_quantity(text):
    words = 0
    for i, char in enumerate(text):
        if char == " " or is_final_punct(char):
            if i > 0 and not is_final_punct(text[i-1]):
                words += 1
    return words


def get_letters_quantity(text):
    letters = 0
    for char in text:
        if char != " " and not is_final_punct(char) and char.isalpha():
            letters += 1
    return letters


def get_sentences_quantity(text):
    sentences = 0
    for char in text:
        if is_final_punct(char):
            sentences += 1
    return sentences


def is_final_punct(char):
    return char in ["!", "?", "."]


main()
